File: java/src/main/java/com/android/widget/LoadingLayout/LoadingLayout.java
package com.android.widget.LoadingLayout;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Animation;
import android.view.animation.RotateAnimation;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.android.java.R;

/**
 * Created by xuzhb on 2020/7/18
 * Desc:加载状态布局
 */
public class LoadingLayout extends LinearLayout {

    //加载状态
    private static final int STATE_LOADING = 0;  //加载中
    private static final int STATE_EMPTY = 1;    //无数据
    private static final int STATE_FAIL = 2;     //加载失败
    private static final int STATE_HIDE = 3;     //隐藏

    private Drawable loadingSrc;     //加载中的图片
    private String loadingDescText;  //加载中的文本描述
    private Drawable emptySrc;       //无数据的图片
    private String emptyDescText;    //无数据的文本描述
    private String emptyActionText;  //无数据时操作按钮的文本
    private Drawable failSrc;        //加载失败的图片
    private String failDescText;     //加载失败的文本描述
    private String failActionText;   //重试的文本描述
    private boolean intercept;       //是否拦截点击事件，为true时底下的View无法收到点击事件

    public void setLoadingSrc(Drawable loadingSrc) {
        this.loadingSrc = loadingSrc;
        mLoadingIv.setImageDrawable(loadingSrc);
    }

    public void setLoadingDescText(String loadingDescText) {
        this.loadingDescText = loadingDescText;
        setDescText(mDescTv, loadingDescText);
    }

    public void setEmptySrc(Drawable emptySrc) {
        this.emptySrc = emptySrc;
        mEmptyIv.setImageDrawable(emptySrc);
    }

    public void setEmptyDescText(String emptyDescText) {
        this.emptyDescText = emptyDescText;
        setDescText(mDescTv, emptyDescText);
    }

    public void setEmptyActionText(String emptyActionText) {
        this.emptyActionText = emptyActionText;
        setActionText(mActionBtn, emptyActionText);
    }

    public void setFailSrc(Drawable failSrc) {
        this.failSrc = failSrc;
        mFailIv.setImageDrawable(failSrc);
    }

    public void setFailDescText(String failDescText) {
        this.failDescText = failDescText;
        setDescText(mDescTv, failDescText);
    }

    public void setFailActionText(String failActionText) {
        this.failActionText = failActionText;
        setActionText(mActionBtn, failActionText);
    }

    //获取根布局，以便设置布局的LayoutParams
    public LinearLayout getRootLayout() {
        return mRootLayout;
    }

    private LinearLayout mRootLayout;
    private ImageView mLoadingIv;
    private ImageView mEmptyIv;
    private ImageView mFailIv;
    private TextView mDescTv;
    private Button mActionBtn;

    public LoadingLayout(Context context) {
        this(context, null);
    }

    public LoadingLayout(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LoadingLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context);
        if (attrs != null) {
            parseAttributes(context, attrs);
        }
        setLoadState(STATE_HIDE);
    }

    private void initView(Context context) {
        View layout = LayoutInflater.from(context).inflate(R.layout.layout_loading, this);
        mRootLayout = layout.findViewById(R.id.root_ll);
        mLoadingIv = layout.findViewById(R.id.loading_iv);
        mEmptyIv = layout.findViewById(R.id.empty_iv);
        mFailIv = layout.findViewById(R.id.fail_iv);
        mDescTv = layout.findViewById(R.id.desc_tv);
        mActionBtn = layout.findViewById(R.id.action_btn);
    }

    //获取自定义属性集
    private void parseAttributes(Context context, AttributeSet attrs) {
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.LoadingLayout);
        loadingSrc = ta.getDrawable(R.styleable.LoadingLayout_loadingSrc);
        if (loadingSrc == null) {
            loadingSrc = getResources().getDrawable(R.drawable.ic_load_loading);
        }
        loadingDescText = ta.getString(R.styleable.LoadingLayout_loadingDescText);
        emptySrc = ta.getDrawable(R.styleable.LoadingLayout_emptySrc);
        if (emptySrc == null) {
            emptySrc = getResources().getDrawable(R.drawable.ic_load_empty);
        }
        emptyDescText = ta.getString(R.styleable.LoadingLayout_emptyDescText);
        emptyActionText = ta.getString(R.styleable.LoadingLayout_emptyActionText);
        failSrc = ta.getDrawable(R.styleable.LoadingLayout_failSrc);
        if (failSrc == null) {
            failSrc = getResources().getDrawable(R.drawable.ic_load_fail);
        }
        failDescText = ta.getString(R.styleable.LoadingLayout_failDescText);
        failActionText = ta.getString(R.styleable.LoadingLayout_failActionText);
        intercept = ta.getBoolean(R.styleable.LoadingLayout_intercept, true);
        ta.recycle();
    }

    //开始加载
    public void loadStart() {
        setLoadState(STATE_LOADING);
    }

    //无数据
    public void loadEmpty() {
        setLoadState(STATE_EMPTY);
    }

    //加载失败
    public void loadFail() {
        setLoadState(STATE_FAIL);
    }

    //加载完成，则隐藏
    public void loadComplete() {
        setLoadState(STATE_HIDE);
    }

    private void setLoadState(int loadState) {
        if (loadState == STATE_HIDE) {
            setVisibility(GONE);
        } else {
            setVisibility(VISIBLE);
            mLoadingIv.clearAnimation();  //取消动画
            mLoadingIv.setVisibility(GONE);
            mEmptyIv.setVisibility(GONE);
            mFailIv.setVisibility(GONE);
            mActionBtn.setVisibility(GONE);
            if (loadState == STATE_LOADING) {
                mLoadingIv.setVisibility(VISIBLE);
                mLoadingIv.setImageDrawable(loadingSrc);
                RotateAnimation animation = new RotateAnimation(0f, 359f,
                        Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
                animation.setInterpolator(new AccelerateDecelerateInterpolator());
                animation.setDuration(800);
                animation.setRepeatCount(Animation.INFINITE);
                animation.setRepeatMode(Animation.RESTART);
                mLoadingIv.startAnimation(animation);
                setDescText(mDescTv, loadingDescText);
            } else if (loadState == STATE_EMPTY) {
                mEmptyIv.setVisibility(VISIBLE);
                mEmptyIv.setImageDrawable(emptySrc);
                setDescText(mDescTv, emptyDescText);
                setActionText(mActionBtn, emptyActionText);
                mActionBtn.setOnClickListener(v -> {
                    if (mOnEmptyListener != null) {
                        mOnEmptyListener.onEmpty();
                    }
                });
            } else if (loadState == STATE_FAIL) {
                mFailIv.setVisibility(VISIBLE);
                mFailIv.setImageDrawable(failSrc);
                setDescText(mDescTv, failDescText);
                setActionText(mActionBtn, failActionText);
                mActionBtn.setOnClickListener(v -> {
                    if (mOnFailListener != null) {
                        mOnFailListener.onFail();
                    }
                });
            }
        }
    }

    private void setDescText(TextView tv, String desc) {
        if (TextUtils.isEmpty(desc)) {
            tv.setVisibility(GONE);
        } else {
            tv.setVisibility(VISIBLE);
            tv.setText(desc);
        }
    }

    private void setActionText(Button btn, String text) {
        if (TextUtils.isEmpty(text)) {
            btn.setVisibility(View.GONE);
        } else {
            btn.setVisibility(View.VISIBLE);
            btn.setText(text);
        }
    }

    public OnEmptyListener mOnEmptyListener;

    //无数据时相应的操作
    public void setOnEmptyListener(OnEmptyListener listener) {
        this.mOnEmptyListener = listener;
    }

    public interface OnEmptyListener {
        void onEmpty();
    }

    public OnFailListener mOnFailListener;

    //加载失败时相应的操作，如点击重试
    public void setOnFailListener(OnFailListener listener) {
        this.mOnFailListener = listener;
    }

    public interface OnFailListener {
        void onFail();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return intercept;  //拦截点击事件
    }
}

File: kotlin/src/main/java/com/android/widget/LoadingLayout/LoadingLayout.kt
package com.android.widget.LoadingLayout

import android.content.Context
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.Animation
import android.view.animation.RotateAnimation
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.AttrRes
import com.android.basicproject.R

/**
 * Created by xuzhb on 2020/7/18
 * Desc:加载状态布局
 */
class LoadingLayout @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, @AttrRes defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    companion object {
        private const val STATE_LOADING = 0  //加载中
        private const val STATE_EMPTY = 1    //无数据
        private const val STATE_FAIL = 2     //加载失败
        private const val STATE_HIDE = 3     //隐藏
    }

    var loadingSrc: Drawable? = null  //加载中的图片
        set(value) {
            field = value
            mLoadingIv.setImageDrawable(value)
        }
    var loadingDescText: String = ""  //加载中的文本描述
        set(value) {
            field = value
            setDescText(mDescTv, value)
        }
    var emptySrc: Drawable? = null  //无数据的图片
        set(value) {
            field = value
            mEmptyIv.setImageDrawable(value)
        }
    var emptyDescText: String = ""  //无数据的文本描述
        set(value) {
            field = value
            setDescText(mDescTv, value)
        }
    var emptyActionText: String = ""  //无数据时操作按钮的文本
        set(value) {
            field = value
            setActionText(mActionBtn, value)
        }
    var failSrc: Drawable? = null  //加载失败的图片
        set(value) {
            field = value
            mFailIv.setImageDrawable(value)
        }
    var failDescText: String = ""  //加载失败的文本描述
        set(value) {
            field = value
            setDescText(mDescTv, value)
        }
    var failActionText: String = ""  //重试的文本描述
        set(value) {
            field = value
            setActionText(mActionBtn, value)
        }

    //获取根布局，以便设置布局的LayoutParams
    fun getRootLayout(): LinearLayout = mRootLayout

    private val mRootLayout: LinearLayout
    private val mLoadingIv: ImageView
    private val mEmptyIv: ImageView
    private val mFailIv: ImageView
    private val mDescTv: TextView
    private val mActionBtn: Button

    init {
        val layout = LayoutInflater.from(context).inflate(R.layout.layout_loading, this)
        with(layout) {
            mRootLayout = findViewById(R.id.root_ll)
            mLoadingIv = findViewById(R.id.loading_iv)
            mEmptyIv = findViewById(R.id.empty_iv)
            mFailIv = findViewById(R.id.fail_iv)
            mDescTv = findViewById(R.id.desc_tv)
            mActionBtn = findViewById(R.id.action_btn)
        }
        attrs?.let {
            val ta = context.obtainStyledAttributes(it, R.styleable.LoadingLayout)
            loadingSrc = ta.getDrawable(R.styleable.LoadingLayout_loadingSrc) ?: resources.getDrawable(R.drawable.ic_load_loading)
            loadingDescText = ta.getString(R.styleable.LoadingLayout_loadingDescText) ?: ""
            emptySrc = ta.getDrawable(R.styleable.LoadingLayout_emptySrc) ?: resources.getDrawable(R.drawable.ic_load_empty)
            emptyDescText = ta.getString(R.styleable.LoadingLayout_emptyDescText) ?: ""
            emptyActionText = ta.getString(R.styleable.LoadingLayout_emptyActionText) ?: ""
            failSrc = ta.getDrawable(R.styleable.LoadingLayout_failSrc) ?: resources.getDrawable(R.drawable.ic_load_fail)
            failDescText = ta.getString(R.styleable.LoadingLayout_failDescText) ?: ""
            failActionText = ta.getString(R.styleable.LoadingLayout_failActionText) ?: ""
            ta.recycle()
        }
//        loadStart()
    }

    //开始加载
    fun loadStart() {
        setLoadState(STATE_LOADING)
    }

    //无数据
    fun loadEmpty() {
        setLoadState(STATE_EMPTY)
    }

    //加载失败
    fun loadFail() {
        setLoadState(STATE_FAIL)
    }

    //加载完成，则隐藏
    fun loadComplete() {
        setLoadState(STATE_HIDE)
    }

    private fun setLoadState(loadState: Int) {
        if (loadState == STATE_HIDE) {
            visibility = View.GONE
        } else {
            visibility = View.VISIBLE
            mLoadingIv.clearAnimation()  //取消动画
            mLoadingIv.visibility = View.GONE
            mEmptyIv.visibility = View.GONE
            mFailIv.visibility = View.GONE
            mActionBtn.visibility = View.GONE
            when (loadState) {
                STATE_LOADING -> {
                    mLoadingIv.visibility = View.VISIBLE
                    mLoadingIv.setImageDrawable(loadingSrc)
                    val animation = RotateAnimation(
                        0f, 359f,
                        Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f
                    )
                    mLoadingIv.startAnimation(animation.apply {
                        duration = 800
                        repeatCount = Animation.INFINITE
                        repeatMode = Animation.RESTART
                        interpolator = AccelerateDecelerateInterpolator()
                    })
                    setDescText(mDescTv, loadingDescText)
                }
                STATE_EMPTY -> {
                    mEmptyIv.visibility = View.VISIBLE
                    mEmptyIv.setImageDrawable(emptySrc)
                    setDescText(mDescTv, emptyDescText)
                    setActionText(mActionBtn, emptyActionText)
                    mActionBtn.setOnClickListener { mOnEmptyListener?.invoke() }
                }
                STATE_FAIL -> {
                    mFailIv.visibility = View.VISIBLE
                    mFailIv.setImageDrawable(failSrc)
                    setDescText(mDescTv, failDescText)
                    setActionText(mActionBtn, failActionText)
                    mActionBtn.setOnClickListener { mOnFailListener?.invoke() }
                }
            }
        }
    }

    private fun setDescText(tv: TextView, desc: String) {
        if (TextUtils.isEmpty(desc)) {
            tv.visibility = View.GONE
        } else {
            tv.visibility = View.VISIBLE
            tv.text = desc
        }
    }

    private fun setActionText(btn: Button, text: String) {
        if (TextUtils.isEmpty(text)) {
            btn.visibility = View.GONE
        } else {
            btn.visibility = View.VISIBLE
            btn.text = text
        }
    }

    private var mOnEmptyListener: (() -> Unit)? = null

    //无数据时相应的操作
    fun setOnEmptyListener(listener: () -> Unit) {
        mOnEmptyListener = listener
    }

    private var mOnFailListener: (() -> Unit)? = null

    //加载失败时相应的操作，如点击重试
    fun setOnFailListener(listener: () -> Unit) {
        mOnFailListener = listener
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        return true  //拦截点击事件
    }

}